yew/src/virtual_dom/mod.rs

```diff
@@ -314,42 +314,78 @@ impl Default for Attributes {
 
 /// A set of classes.
 #[derive(Debug, Clone, Default)]
-pub struct Classes {
-    set: IndexSet<String>,
+pub struct Classes<T = String> {
```

Comment:
Not sure if we should really make `Classes` generic.
What do you think about using `Cow<'static, str>`?
It certainly has a bit more overhead but it also allows mixing owned with borrowed in cases like this:
```rust
let mut classes = Classes::default();
classes.push("my-static-class");
classes.push(format!("is-size-{}", self.size));
```

I'm not entirely convinced that this is the best solution either but it's a big step up from using `String` like before.

Reply:
I've never used Cow but it sounds like a great idea!

I'm not too fan of generics either but it did the job pretty well here.

Reply:
I pushed something but don't merge. I don't think it works

Reply:
I don't know how to make a Classes that will have a 'static str. I don't think it's possible with `From` as there are conflicting implementations.

```rust
impl From<&'static str> for Classes {
    fn from(t: &str) -> Self {
        let set = t
            .split_whitespace()
            .filter(|c| !c.is_empty())
            .collect();
        Self { set }
    }
}

impl From<&str> for Classes {
    fn from(t: &str) -> Self {
        let set = t
            .split_whitespace()
            .filter(|c| !c.is_empty())
            .map(|x| x.to_string().into())
            .collect();
        Self { set }
    }
}
```

I will have to create a method push_static which would make a Cow::Borrowed (same problem for extend).

I revert to the impl with generic?

Reply:
Yeah, until Rust finally blesses us with trait specialization we can't have both.
But I don't think that's a problem. Just implement `From<String>` and `From<&'static str>` because non static string slices need to be cloned anyway and I believe it's better to be upfront about that

Reply:
For the `Extend` trait I would still use `Into<Self>` for the item type. An implicit contract of `Classes` is that `classes.push("multiple classes")` is handled properly (that's why we split on whitespace) and the current `Extend` doesn't uphold this.
Also, using `Into<Self>` handles both `String` and `&'static str` in a way that doesn't require cloning in the case of `&'static str`

Reply:
> Also, using `Into<Self>` handles both `String` and `&'static str` in a way that doesn't require cloning in the case of `&'static str`

I am pretty sure it is allocating because it uses From<&str> which uses to_string

Reply:
That's what my first comment is about :)

In `From<&'static str>` we don't have to use `to_string()` we can just do this:
```rust
 impl From<&'static str> for Classes {
    fn from(t: &'static str) -> Self {
        let set = t
            .split_whitespace()
            .map(Cow::borrowed)
            .collect();
        Self { set }
    }
}
```

Reply:
> In `From<&'static str>` we don't have to use `to_string()` we can just do this:

Sorry I am not on my PC. But what about not static &str? That won't work anymore

Reply:
Yes, that's true. I'm arguing that that's not really a problem.
I don't think `&str` is a very common type for this application anyway. You either use a literal string (`"my-class"`) or you build an owned `String` eg. using `format!("my-class-{}", self.class_mod)`.
In rare cases you might take reference to a string from somewhere else but in those cases I believe it's reasonable for the user to call `.clone()` or `.to_owned()` themselves.
In my opinion implicitly cloning data is seldom the best approach. It can lead users to believe that an operation is less expensive than it actually is (there are exceptions for convenience, of course).

If we implement it for `&str` instead of `&'static str` there's really no point in using `Cow` in the first place since we have to clone the strings anyway.

Reply:
If you're worried about it, I wouldn't mind giving  `Classes` a generic lifetime:
```rust
struct Classes<'a> {
  set: IndexSet<Cow<'a, str>>
}

impl<'a> From<&'a str> for Classes<'a> {
  fn from(v: &'a str) -> Self {
    let set = v.split_whitespace().map(Cow::Borrowed).collect(); 
    Self { set }
  }
}
```

Since `Classes` is only used temporarily it doesn't matter if it's short-lived so you can use `&str` in the `html!` macro.
This doesn't change the situation for components though. In fact, we would have to explicitly add the `'static` lifetime:
```rust
struct MyProps {
  pub class: Classes<'static>,
}
```

Because of this (and my previous argument that `&str` is the exception) I think we should stick to `'static`. What do you think?

Reply:
Ah yes sure! My point was just about breaking the API: not allowing not-static &str anymore. But if you're find with it I will break it. I prefer 'static over 'a because lifetimes with yew doesn't really work anyway.

> What do you think?

I think composing strings in classes is never what you want because CSS is never generated dynamically while on a page.

My personal opinion is that String Classes shouldn't even be allowed at all as it allows bad pattern. (So I believe &'static str-only is better than Cow or String.)

Though we should let the developer use a String in a class if they really want to:

```rust
let my_custom_classes = format!("...", ...);

html! {
    <div class=my_custom_classes />
}
```

But NOT with the Classes itself:

```rust
let my_custom_classes = Classes::from(format!("...", ...)); // <--- invalid
```

Reply:
(If you don't mind checking the logs, I'm a bit stuck with a recursion :sweat_smile: )

Reply:
I fixed it using Clone but I'm unsure if it's good...

Reply:
> My personal opinion is that String Classes shouldn't even be allowed at all as it allows bad pattern. (So I believe &'static str-only is better than Cow or String.)

You have a point but there are a few cases where `String`s are useful or even required. `Classes` needs to support `From<String>` because it's what the `html!` macro uses. If `Classes` didn't support it, the example you gave wouldn't compile:
```rust
let my_custom_classes = format!("...", ...);

html! {
    <div class=my_custom_classes />
}
```

Anyway, I think we're on the right track here.